src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp

```diff
@@ -0,0 +1,584 @@
+   * @param[out] receiverNodeIds indices of the nodes of the element where the receiver is located
+   * @param[out] receiverConstants constant part of the receiver term
+   * @param[out] sourceValue value of the temporal source (eg. Ricker)
+   * @param[in] dt time-step
+   * @param[in] timeSourceFrequency the central frequency of the source
+   * @param[in] rickerOrder order of the Ricker wavelet
+   */
+  template< typename EXEC_POLICY, typename FE_TYPE >
+  static void
+  launch( localIndex const size,
+          localIndex const numNodesPerElem,
+          localIndex const numFacesPerElem,
+          arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const X,
+          arrayView1d< integer const > const elemGhostRank,
+          arrayView2d< localIndex const, cells::NODE_MAP_USD > const & elemsToNodes,
+          arrayView2d< localIndex const > const elemsToFaces,
+          arrayView2d< real64 const > const & elemCenter,
+          arrayView2d< real64 const > const faceNormal,
+          arrayView2d< real64 const > const faceCenter,
+          arrayView2d< real64 const > const sourceCoordinates,
+          arrayView1d< localIndex > const sourceIsAccessible,
+          arrayView1d< localIndex > const sourceElem,
+          arrayView2d< localIndex > const sourceNodeIds,
+          arrayView2d< real64 > const sourceConstants,
+          arrayView2d< real64 const > const receiverCoordinates,
+          arrayView1d< localIndex > const receiverIsLocal,
+          arrayView1d< localIndex > const rcvElem,
+          arrayView2d< localIndex > const receiverNodeIds,
+          arrayView2d< real64 > const receiverConstants,
+          arrayView2d< real32 > const sourceValue,
+          real64 const dt,
+          real32 const timeSourceFrequency,
+          localIndex const rickerOrder )
+  {
+
+    forAll< EXEC_POLICY >( size, [=] GEOSX_HOST_DEVICE ( localIndex const k )
+    {
+      real64 const center[3] = { elemCenter[k][0],
+                                 elemCenter[k][1],
+                                 elemCenter[k][2] };
+
+      // Step 1: locate the sources, and precompute the source term
+
+      /// loop over all the source that haven't been found yet
+      for( localIndex isrc = 0; isrc < sourceCoordinates.size( 0 ); ++isrc )
+      {
+        if( sourceIsAccessible[isrc] == 0 )
+        {
+          real64 const coords[3] = { sourceCoordinates[isrc][0],
+                                     sourceCoordinates[isrc][1],
+                                     sourceCoordinates[isrc][2] };
+
+          bool const sourceFound =
+            WaveSolverUtils::locateSourceElement( numFacesPerElem,
+                                                  center,
+                                                  faceNormal,
+                                                  faceCenter,
+                                                  elemsToFaces[k],
+                                                  coords );
```

Comment:
In a later PR, we should remove this function and use the one available in `ComputationalGeometry.hpp` to benefit from the improvements made by @untereiner and @dkachuma